### ISSUE TITLE
Training on new Dataset

### ISSUE THREAD

**Issue by sunline059** [AlexeyAB/darknet#51]
Is it possible to train Yolo v2 on my own dataset while using e.g. yolo-voc.weights as pre trained model?
Like if i want to train yolo to detect my own dog in a video and train it with images of my dog. Since the voc model already can detect dogs pretty well, is it possible to use these pre trained weights to gain improved or sped-up training results on my own dataset with 1 class or do we always have to start from scratch or use darknet.23  ?

**Comment by AlexeyAB** [comment]
@sunline059 

* The best way is to use `darknet19_448.conv.23` convolutional weights.

* If you want to use `yolo-voc.weights` to fine tuning, then you should use the same `yolo-voc.cfg`-file with the same number of classes (20 classes), where your dog will be the same class number 11.

* Also you can do `./darknet partial darknet19_448.cfg yolo-voc.weights yolo-voc.conv.23 23`
  And then train your custom model but withyolo-voc.conv.23 instead of darknet19_448.conv.23:
  `./darknet detector train data/obj.data yolo-obj.cfg yolo-voc.conv.23` as said here: https://github.com/AlexeyAB/darknet#how-to-train-to-detect-your-custom-objects

----

In your case you can simply detect only dogs by using `yolo-voc.cfg` and `yolo-voc.weights` without any training. 
Simply change this line from `if(prob > thresh){` to `if(prob > thresh && class == 11){` 
https://github.com/AlexeyAB/darknet/blob/d360b6a42a91eef66b0085c81ecbff4792986402/src/image.c#L187

**Comment by sunline059** [comment]
@AlexeyAB 

Thank you for the quick reply and clarification.
So in general if i want to train my own images i have to always start with darknet19_448.conv.23 ?
If i want to add my own images to voc i would have to train from scratch with the voc dataset and add my images to that ? Would it work if i just take the dog images from voc, add my own photos and train from darknet19_448.conv.23 ?

I think i didn't really understand what the difference with the pre trained weight files and the ready to use voc or coco weights is. Can you tell me in short how they are different? I thought that darknet19_448.conv.23 is pre trained on 100 classes of image net and voc weights are trained on the voc challenge dataset and classes. So i don't really understand why i can't use voc weights as "pre trained" weights to learn my own classes. 

Maybe you have some links to good literature to guide me on that pre training and training stuff ?

Thanks in advance!

**Comment by AlexeyAB** [comment]
> So in general if i want to train my own images i have to always start with darknet19_448.conv.23 ?

Yes.

> If i want to add my own images to voc i would have to train from scratch with the voc dataset and add my images to that ?

Yes.

> Would it work if i just take the dog images from voc, add my own photos and train from darknet19_448.conv.23 ?

Yes, but you should class number of dog in voc.

> I thought that darknet19_448.conv.23 is pre trained on 100 classes of image net and voc weights are trained on the voc challenge dataset and classes. 

Not 100, but 1000 classes  and 1.4 millions images in the ImageNet **classification** dataset.

>  So i don't really understand why i can't use voc weights as "pre trained" weights to learn my own classes.

You can use voc weights as "pre trained" weights.

* if you want to train by using `yolo-voc.weights` (**with output layer**) then your dog should have the same 11 class number. Also it can leads to overfitting due that yolo-voc.weights already trained at 45 000 iterations, and you'll train it more.

* if you want to train by using `yolo-voc.conv.23` (**without output layer**)  then it should not give you many advantages compared with darknet19_448.conv.23 - this weights already pre-trained on ImageNet 1000 classes, where are dogs.

**Comment by sunline059** [comment]
Thanks for the clarification, i think i got it now.

The pre trained networks like darknet19_448.conv.23 are trained but are missing the output layers. 
yolo-voc has output layers and was trained for 45000 iterations so i would probably get overfitting if i train it more.
And we can't use yolo-voc to train from scratch because the output layers are already trained - therefore i woukd have to use a voc weights without output layers which is not much different to darknet19_448.conv.23. Does that sound about right ?

**Comment by AlexeyAB** [comment]
Yes everything is correct.

But this `voc weights without output layers which is not much different to darknet19_448.conv.23` is only my guess, maybe it will give you better results than darknet19_448.conv.23.

You can try to train:

* extract only conv-weights `yolo-voc.conv.23` from VOC `./darknet partial darknet19_448.cfg yolo-voc.weights yolo-voc.conv.23 23`

* train on your custom data by using `./darknet detector train data/obj.data yolo-obj.cfg yolo-voc.conv.23`

**Comment by sunline059** [comment]
Ok, then i will test that out and see what results i get.
Just two general question: How would you go about if you want to train differnet species of dogs but still keep the identifier "dog" for unknown species ? And does it make sense to also add images to the training with just background and no bounding boxes ?

**Comment by AlexeyAB** [comment]
>  And does it make sense to also add images to the training with just background and no bounding boxes ?

Yes, this improves accuracy.

> Just two general question: How would you go about if you want to train differnet species of dogs but still keep the identifier "dog" for unknown species ?

* A simple way is to mark out all known dogs as separate classes, and all other dogs as another one separate class.

* A more accurate solution is to use the Yolo 9000: https://github.com/AlexeyAB/darknet/issues/45#issuecomment-288488580

**Comment by sunline059** [comment]
Thanks again Alexey you got me right on track and also really nice job you are doing with your windows build, really appreciate it!
